### ISSUE TITLE
Fault Tolerant Deployment

### ISSUE THREAD

**Issue by xavipanda** [nginxinc/nginx-service-mesh#19]
Hi,

This is not described anywhere in the documentation, but imagine there was a need to implement a high-performant fully-redundant solution with NGiNX Service Mesh:
* What would be the patterns to follow
* What would be the changes between the standard deployment and this one
* There is any way to avoid nginx-meshctl on the provisioning process ?

Appreciate feedback regarding this.
Thank you

**Comment by f5yacobucci** [comment]
The current pattern we employ (and have verified through testing) is a stateless, fail-fast approach. Each controller and webhook rebuilds world view on startup and will actuate configurations; from a level set approach, once synced at startup; reacting to changes and events as they arrive from the Kubernetes API server.

For small to medium sized applications this has served well. Our control plane architecture will continue to mature, but horizontal scalability is unsupported currently.

The sole method for provisioning is via nginx-meshctl. We have work items identified to support Helm, and also convenience functions which will output Kubernetes configuration yaml/json files for users to consume and examine at will.

**Comment by xavipanda** [comment]
I understand. Let me maybe rephrase my question:
- Can we have multiple replicas (in the same cluster for) mesh-api mesh-metrics  pods ? 

If there is only 1 pod for spire-server, mesh-api, mesh-metrics... definitively is a SPOF. So the question is mostly about can we scale these safely without producing race/misbehaviours among them ? @malanmurphy  thanks

**Comment by f5yacobucci** [comment]
Not at the moment with the `nginx-mesh-api` controllers. We have not tested `nginx-mesh-metrics` though. Under the hood we're using APIService extensions. I'll need to look into if API extensions can be made to support replicas or DaemonSets.

**Comment by xavipanda** [comment]
I got around some changes for the Chart, and i'll submit them soon. Mostly its about the scenario (tested) where spire-server is using SQL as datastore with multiple replicas.
In the case of nginx-mesh-metrics, the API Extension allows multiple pods (service->endpoints).
For the nginx-mesh-api controller, that would depends of how the controller was designed. By example, the runtime-controller allows leader election and IMHO shouldn't be difficult to accomplish if the source code becomes open.

**Comment by xavipanda** [comment]
This is achieved mostly by improving the helm chart and making spire HA, thru remote datastore ex. rds.
the piece of HA on the controller has to be done by the dev team.
